using_restricted_array.rb

```diff
@@ -7,41 +7,88 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+    i = 0
+    until array[i] == nil
+      puts array[i]
+      i += 1
+    end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  until array[length - 1] < 0
+    return true if array[length - 1] == value_to_find
+    length -= 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  max = array[0]
+  length = length - 1
+  until length == 0
+    if array[length] > max
+      max = array[length]
+    end
+    length -= 1
+  end
+  return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  min = array[0]
+  length = length - 1
+  until length == 0
+    min = array[length - 1] if array[length - 1] < min
```

Comment:
You are not currently checking array[length -1] i.e. the last element in the array. Since on line 54, length is reduced by one and in the loop you are checking element at index length-1.